### DISCUSSION TITLE
Using GIO objects

### DISCUSSION THREAD

**Discussion by hatahet** [gstreamer-java/gst1-java-core#258] [General]
I'm trying to implement a callback for `hlssink2`'s [`get-fragment-stream`](https://gstreamer.freedesktop.org/documentation/hls/hlssink2.html?gi-language=python#hlssink2::get-fragment-stream) signal. The callback is documented to return a `Gio.OutputStream` object. It seems that the object is not exposed in this library. Is this assessment correct? If so, what workarounds are possible at this time?

**Comment by neilcsmith-net** [comment]
You can create your own types by extending `GObject` and registering a `NativeObject.TypeProvider` in a way that will be picked up by `ServiceLoader`.  See code at https://github.com/gstreamer-java/gst1-java-core/blob/master/src/org/freedesktop/gstreamer/Gst.java#L647

Next release should make `GObject` non-abstract so allow mapping just to that.

I also have some ideas for allowing custom callback registration to make that easier to do.

**Reply by hatahet** [reply]
Thank you @neilcsmith-net, I'll look into that. I was also using a `format-location` signal callback for [`splitmuxsink`](https://gstreamer.freedesktop.org/documentation/multifile/splitmuxsink.html?gi-language=python#splitmuxsink::format-location), and I noticed that if I invoke the GC, the program would crash after a couple of times. I'm suspecting that Java's `String` class does not conform to the return type that is required here. Would I also need to implement a custom `gchar` native object type provider to properly use as a return type for this callback?

**Reply by neilcsmith-net** [reply]
A quick look does suggest that's the likely issue. I think you might need to follow the advice in the docs there and use `g_strdup`, which we don't seem to have mapped even in low level.  Probably as `Pointer g_strdup(String str)`.
